paperqa/agents/docs.py

```diff
@@ -0,0 +1,282 @@
+
+from anthropic import AsyncAnthropic
+from openai import AsyncOpenAI
+
+from .. import (
+    Answer,
+    AnthropicLLMModel,
+    Docs,
+    OpenAILLMModel,
+    embedding_model_factory,
+    llm_model_factory,
+)
+from ..llms import LangchainLLMModel
+from .models import (
+    QueryRequest,
+)
+
+USE_VERTEX = True
+try:
+    from langchain_google_vertexai import ChatVertexAI, HarmBlockThreshold, HarmCategory
+except ImportError:
+    USE_VERTEX = False
+
+logger = logging.getLogger(__name__)
+
+
+# Index 0 is for prompt tokens, index 1 is for completion tokens
+costs: dict[str, tuple[float, float]] = {
+    "claude-2": (11.02 / 10**6, 32.68 / 10**6),
+    "claude-instant-1": (1.63 / 10**6, 5.51 / 10**6),
+    "claude-3-sonnet-20240229": (3 / 10**6, 15 / 10**6),
+    "claude-3-5-sonnet-20240620": (3 / 10**6, 15 / 10**6),
+    "claude-3-opus-20240229": (15 / 10**6, 75 / 10**6),
+    "babbage-002": (0.0004 / 10**3, 0.0004 / 10**3),
+    "gpt-3.5-turbo": (0.0010 / 10**3, 0.0020 / 10**3),
+    "gpt-3.5-turbo-1106": (0.0010 / 10**3, 0.0020 / 10**3),
+    "gpt-3.5-turbo-0613": (0.0010 / 10**3, 0.0020 / 10**3),
+    "gpt-3.5-turbo-0301": (0.0010 / 10**3, 0.0020 / 10**3),
+    "gpt-3.5-turbo-0125": (0.0005 / 10**3, 0.0015 / 10**3),
+    "gpt-4-1106-preview": (0.010 / 10**3, 0.030 / 10**3),
+    "gpt-4-0125-preview": (0.010 / 10**3, 0.030 / 10**3),
+    "gpt-4-turbo-2024-04-09": (10 / 10**6, 30 / 10**6),
+    "gpt-4-turbo": (10 / 10**6, 30 / 10**6),
+    "gpt-4": (0.03 / 10**3, 0.06 / 10**3),
+    "gpt-4-0613": (0.03 / 10**3, 0.06 / 10**3),
+    "gpt-4-0314": (0.03 / 10**3, 0.06 / 10**3),
+    "gpt-4o": (2.5 / 10**6, 10 / 10**6),
+    "gpt-4o-2024-05-13": (5 / 10**6, 15 / 10**6),
+    "gpt-4o-2024-08-06": (2.5 / 10**6, 10 / 10**6),
+    "gpt-4o-mini": (0.15 / 10**6, 0.60 / 10**6),
+    "gemini-1.5-flash": (0.35 / 10**6, 0.35 / 10**6),
+    "gemini-1.5-pro": (3.5 / 10**6, 10.5 / 10**6),
+    # supported Anyscale models per
+    # https://docs.anyscale.com/endpoints/text-generation/query-a-model
+    "meta-llama/Meta-Llama-3-8B-Instruct": (0.15 / 10**6, 0.15 / 10**6),
+    "meta-llama/Meta-Llama-3-70B-Instruct": (1.0 / 10**6, 1.0 / 10**6),
+    "mistralai/Mistral-7B-Instruct-v0.1": (0.15 / 10**6, 0.15 / 10**6),
+    "mistralai/Mixtral-8x7B-Instruct-v0.1": (1.0 / 10**6, 1.0 / 10**6),
+    "mistralai/Mixtral-8x22B-Instruct-v0.1": (1.0 / 10**6, 1.0 / 10**6),
+}
+
+
+def get_llm_name(llm: Any) -> str:
+    if isinstance(llm, str):
+        return llm
+    try:
+        return llm.model_name
+    except AttributeError:
+        return llm.model
+
+
+def get_cost(model: str, tokens: int, is_completion: bool) -> float:
+    if model in costs:  # Prefer our internal costs model
+        model_costs: tuple[float, float] = costs[model]
+    else:
+        logger.warning(f"Model {model} not found in costs.")
+        return 0.0
+    return tokens * model_costs[int(is_completion)]
+
+
+def compute_cost(token_counts: dict[str, list[int]]) -> float:
```

Comment:
The names `get_cost` and `compute_cost` are so similar, but they do different things. Can we rename them to be more intuitive:
- `get_cost` to `compute_cost`
- `compute_cost` to `bulk_compute_cost`

Reply:
Went with `compute_total_model_token_cost` and `compute_model_token_cost`